File: src/sdk/typescript/src/feature.ts
// 本ファイルは k1s0 TypeScript SDK の Feature 動詞統一 facade（評価部のみ）。
import { createPromiseClient } from "@connectrpc/connect";
import type { K1s0Client } from "./client.js";
import { FeatureService } from "./proto/k1s0/tier1/feature/v1/feature_service_connect.js";
import type { FlagMetadata } from "./proto/k1s0/tier1/feature/v1/feature_service_pb.js";

/** FeatureFacade は FeatureService の動詞統一 facade。 */
export class FeatureFacade {
  private readonly client: K1s0Client;

  constructor(client: K1s0Client) {
    this.client = client;
  }

  private buildReq(flagKey: string, evalCtx: Record<string, string>) {
    return {
      flagKey,
      evaluationContext: evalCtx,
      context: this.client.tenantContext(),
    };
  }

  /** evaluateBoolean は boolean Flag 評価。 */
  async evaluateBoolean(flagKey: string, evalCtx: Record<string, string> = {}): Promise<{ value: boolean; metadata?: FlagMetadata }> {
    const raw = createPromiseClient(FeatureService, this.client.transport);
    const resp = await raw.evaluateBoolean(this.buildReq(flagKey, evalCtx));
    return { value: resp.value, metadata: resp.metadata };
  }

  /** evaluateString は string Flag 評価。 */
  async evaluateString(flagKey: string, evalCtx: Record<string, string> = {}): Promise<{ value: string; metadata?: FlagMetadata }> {
    const raw = createPromiseClient(FeatureService, this.client.transport);
    const resp = await raw.evaluateString(this.buildReq(flagKey, evalCtx));
    return { value: resp.value, metadata: resp.metadata };
  }

  /** evaluateNumber は number Flag 評価。 */
  async evaluateNumber(flagKey: string, evalCtx: Record<string, string> = {}): Promise<{ value: number; metadata?: FlagMetadata }> {
    const raw = createPromiseClient(FeatureService, this.client.transport);
    const resp = await raw.evaluateNumber(this.buildReq(flagKey, evalCtx));
    return { value: resp.value, metadata: resp.metadata };
  }

  /** evaluateObject は object Flag 評価（JSON シリアライズ済 bytes）。 */
  async evaluateObject(flagKey: string, evalCtx: Record<string, string> = {}): Promise<{ valueJson: Uint8Array; metadata?: FlagMetadata }> {
    const raw = createPromiseClient(FeatureService, this.client.transport);
    const resp = await raw.evaluateObject(this.buildReq(flagKey, evalCtx));
    return { valueJson: resp.valueJson, metadata: resp.metadata };
  }
}

File: src/sdk/rust/crates/k1s0-sdk/src/feature.rs
// 本ファイルは k1s0-sdk の Feature 動詞統一 facade（評価部のみ）。
use crate::client::Client;
use crate::proto::k1s0::tier1::feature::v1::{
    feature_service_client::FeatureServiceClient, EvaluateRequest, FlagMetadata,
};
use std::collections::HashMap;
use tonic::{transport::Channel, Status};

/// FeatureFacade は FeatureService の動詞統一 facade。
pub struct FeatureFacade {
    client: Client,
    raw: FeatureServiceClient<Channel>,
}

impl FeatureFacade {
    pub(crate) fn new(client: Client) -> Self {
        let raw = FeatureServiceClient::new(client.channel());
        Self { client, raw }
    }

    fn make_req(&self, flag_key: &str, evaluation_context: HashMap<String, String>) -> EvaluateRequest {
        EvaluateRequest {
            flag_key: flag_key.to_string(),
            evaluation_context,
            context: Some(self.client.tenant_context()),
        }
    }

    /// evaluate_boolean は boolean Flag 評価。返り値は (value, metadata)。
    pub async fn evaluate_boolean(
        &mut self,
        flag_key: &str,
        eval_ctx: HashMap<String, String>,
    ) -> Result<(bool, Option<FlagMetadata>), Status> {
        let resp = self.raw.evaluate_boolean(self.make_req(flag_key, eval_ctx)).await?.into_inner();
        Ok((resp.value, resp.metadata))
    }

    /// evaluate_string は string Flag 評価。
    pub async fn evaluate_string(
        &mut self,
        flag_key: &str,
        eval_ctx: HashMap<String, String>,
    ) -> Result<(String, Option<FlagMetadata>), Status> {
        let resp = self.raw.evaluate_string(self.make_req(flag_key, eval_ctx)).await?.into_inner();
        Ok((resp.value, resp.metadata))
    }

    /// evaluate_number は number Flag 評価。
    pub async fn evaluate_number(
        &mut self,
        flag_key: &str,
        eval_ctx: HashMap<String, String>,
    ) -> Result<(f64, Option<FlagMetadata>), Status> {
        let resp = self.raw.evaluate_number(self.make_req(flag_key, eval_ctx)).await?.into_inner();
        Ok((resp.value, resp.metadata))
    }

    /// evaluate_object は object Flag 評価（JSON シリアライズ済 bytes）。
    pub async fn evaluate_object(
        &mut self,
        flag_key: &str,
        eval_ctx: HashMap<String, String>,
    ) -> Result<(Vec<u8>, Option<FlagMetadata>), Status> {
        let resp = self.raw.evaluate_object(self.make_req(flag_key, eval_ctx)).await?.into_inner();
        Ok((resp.value_json, resp.metadata))
    }
}
